src/linker/Linker/Driver.cs

```diff
@@ -335,6 +335,12 @@ protected int SetupContext (ILogger customLogger = null)
 
 						continue;
 
+					case "--keep-serialization":
```

Comment:
What is the thinking here? What do you need the explicit option for?

I'm wondering if we could do this differently and make all serialization logic dependant on some internal feature flag which would be set automatically or explicitly if something goes wrong. This would help us a lot with size because we could automatically remove also other serialization attributes and we already have a mechanism via `ILLink.LinkAttributes.xml` to do that.

Reply:
The thinking is that the heuristics are optional in case some scenarios don't want the unpredictable behavior or size impact, or want to deal with specific serialized types manually.

> make all serialization logic dependant on some internal feature flag which would be set automatically or explicitly if something goes wrong

If by "something goes wrong" you mean the linker misses some types that are serialized by the app, I don't see how we could tell. Or do you mean something like: if we see a call to `XmlSerializer.ctor` we activate the serialization logic?

Reply:
>  in case some scenarios don't want the unpredictable behavior or size impact

I think we should use the existing feature switch mechanism for that

> Or do you mean something like: if we see a call to XmlSerializer.ctor we activate the serialization logic?

Correct for example someone having serialization logic inside generated SRE code, basically some odd corner cases.

Reply:
I don't think this should be a feature switch since it doesn't match the existing pattern:
- There won't be a runtime feature switch for this so there won't be a predictable way to turn off support for these serializers (unless there are plans I'm not aware of)
- Turning this on on just makes it more likely to work but doesn't guarantee it (so a switch like "XmlSerializer.IsSupported" would be misleading)

Reply:
> There won't be a runtime feature switch for this 

Why not? I'd love to have it to be able to conditional serialization dependant code and attributes with it.

Reply:
Hm, not clear on what value the feature switch is providing. If you want to remove XmlSerializer from your app it seems like the best way to do that is actually removing references to it. The feature switch seems more likely to cause problems instead of solve them.

Reply:
With the existing pattern, I guess a feature switch would just make all of the serializer APIs throw or nop when called, right? I don't think that's a great experience. I could see value in a "feature switch" that gives build-time errors for calls to a serializer, and allows the linker to predictably remove the APIs, but that's another discussion... ;)

That aside, in this change we just need a way to turn off the heuristics. My thinking is that even if we add some kind of feature switch, it should at least be possible to turn off the heuristics when the serializer feature is enabled - that's why I'd prefer it to be a separate argument.

Reply:
The suggestion for feature switch was a conceptual thing, in the same way, we are keeping weak serialization dependencies we could remove them but we don't have infrastructure in linker to do that yet so this will have to wait.

> That aside, in this change we just need a way to turn off the heuristics.

Then let's use the option name which reflects that. Considering we want to have the heuristic enabled by default we could add something like `--disable-serialization-discovery`